### PR TITLE
946 itl mappers

### PR DESCRIPTION
**Pull request by AnneONS**
## Pull Request submission

*Insert detailed bullet points about your changes here!*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `requirements.txt`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/dev_config.yaml [hunk 1]
```diff
@@ -57,11 +57,14 @@ outputs_paths:
   folder: "10_outputs"
   #TODO: add all the output subpaths
   outputs_master: ""
+pnp_paths:
```
Comment (zorge69) [inline src/dev_config.yaml]: I understand where the PNP stuff comes from, still, ideally, we want only one feature to be covered by one branch/pull request.

File: src/mapping/cellno_mapping.py [hunk 2]
```diff
@@ -37,7 +40,7 @@ def clean_validate_cellno_mapper(cellno_df: pd.DataFrame, num: int) -> pd.DataFr
 
 
 def validate_join_cellno_mapper(
-    df: pd.DataFrame, cellno_df: pd.DataFrame, config: dict
+    responses: Tuple[pd.DataFrame, pd.DataFrame], cellno_df: pd.DataFrame, config: dict
```
Comment (zorge69) [inline src/mapping/cellno_mapping.py:37]: It says the argument is a Tuple of two dfs. Yet, the doc string below says the argument is a dataframe.

What is the whole purpose of passing a Tuple? Would it be easier to pass gb_df and ni_df separately? We only have four input arguments.

Still, this is not critical, and i will be happy to pass it anyway, Just please fix the docstring and remove the commented out "num" and get rid of it in the config.

File: src/mapping/cellno_mapping.py [hunk 3]
```diff
@@ -18,13 +20,14 @@ def clean_validate_cellno_mapper(cellno_df: pd.DataFrame, num: int) -> pd.DataFr
         ValueError: If the number of cellnumbers is not as expected.
         ValueError: If the cellnumbers are not in the expected range.
     """
-    # check for unique cellnumbers
     check_mapping_unique(cellno_df, "cell_no")
+
+    # TODO: find out from GZ if we can scrap this.
```
Comment (zorge69) [inline src/mapping/cellno_mapping.py]: Yes, please remove the commented-out area and remove the "num" from line 56, and from config["estimation"]["num_expected_cellnos"]

File: src/mapping/cellno_mapping.py
Comment (zorge69) [inline src/mapping/cellno_mapping.py:1]: This branch is about ITL. It's preferable to edit cellno mapping in a separate branch, unless absolutely critical.

I am happy to pass it.

File: src/mapping/itl_mapping.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+"""Code to join the ITL regions onto the full dataframe using the mapper provided."""
+import pandas as pd
+from typing import Tuple
+
+from src.mapping.mapping_helpers import join_with_null_check
+
```
Comment (zorge69) [inline src/mapping/itl_mapping.py:6]: I think there are too many empty lines, but my flake8 isn't working. Please check linting.

File: src/mapping/itl_mapping.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    responses: Tuple[pd.DataFrame, pd.DataFrame],
+    postcode_mapper: pd.DataFrame,
+    itl_mapper: pd.DataFrame,
+    is_ni: bool = False,
+    pc_col: str = "postcodes_harmonised",
+):
+    """Joins the itl regions onto the full dataframe using the mapper provided
+
+    Args:
+        df (pd.DataFrame): Full dataframe
```
Comment (zorge69) [inline src/mapping/itl_mapping.py]: Fix the docstring to reflect that input is a Tuple

File: src/mapping/itl_mapping.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    Returns:
+        Tuple(gb_df: pd.DataFrame, ni_df: pd.DataFrame): dfs with the ITL regions joined
+    """
+    gb_df, ni_df = responses
+    # first create itl column
+    postcode_mapper = postcode_mapper.rename(columns={"pcd2": pc_col})
+    gb_df = join_with_null_check(gb_df, postcode_mapper, "postcode mapper", pc_col)
+
+    # next join the itl mapper to add the region columns
+    geo_cols = ["LAU121CD", "ITL221CD", "ITL221NM", "ITL121CD", "ITL121NM"]
```
Comment (zorge69) [inline src/mapping/itl_mapping.py:31]: Not critical, but worth checking: shall we put geo_cols = ["LAU121CD", "ITL221CD", "ITL221NM", "ITL121CD", "ITL121NM"] and ni_df["itl"] = "N92000002" into the developer_config? I think it's a good practice to have codes accessible to the users. Please don't do it now, but if you think it's a good idea create a tech debt Jira issue.

File: src/mapping/itl_mapping.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
+"""Code to join the ITL regions onto the full dataframe using the mapper provided."""
```
Comment (zorge69) [inline src/mapping/itl_mapping.py:1]: Please fix the doc stings, all the rest is good.

File: src/mapping/itl_mapping.py (hunk 6 above)
Comment (AnneONS) [reply src/mapping/itl_mapping.py:31]: created ticket [RDRP-958](https://jira.ons.gov.uk/browse/RDRP-958)

**Review by zorge69 — APPROVED** [review]
Checked the outputs 2023_output_gb_sas_24-07-30_v16.csv. All matches, except: some data loss in ITL1 and 2. It comes from records with blank reg_code - bad postcodes?

**Pull request merged**